onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -654,7 +717,7 @@ def create_calibrator(model,
         return MinMaxCalibrater(model, op_types_to_calibrate, augmented_model_path)
     elif calibrate_method == CalibrationMethod.Entropy:
         # default settings for entropy algorithm
-        num_bins = 128 if 'num_bins' not in extra_options else extra_options['num_bins']
+        num_bins = 2048 if 'num_bins' not in extra_options else extra_options['num_bins']
```

Comment:
You can set extra_options['num_bins'] to other value to make num_bins equal to that value

Reply:
@chilo-ms I changed `num_bins` default value because I did't saw histogram's bins become larger in my case, even if I have tried on different size of calibration dataset. 
Base on your latest reply, I believed this feature must pass all necessary test cases before merge into master branch. 
So, let's review next feature.

> No, num_bins = hist.size is not always 128 even though we explicitly set num_bins to 128, it's due to histogram might expand during histogram merging, so hist.size may grow and you will get larger num_bins.

Another code snippet I want to discuss with you is the value `num_merged_bins` inside `get_entropy_threshold` method.
```
...
for i in range(num_half_quantized_bin, zero_bin_index + 1, 1):
            ...
            # quantize p.size bins into quantized bins (default 128 bins)
            quantized_bins = np.zeros(num_quantized_bins, dtype=np.int64)
            
            num_merged_bins = sliced_distribution.size // num_quantized_bins

            # merge bins into quantized bins
            for index in range(num_quantized_bins):
                start = index * num_merged_bins
                end = start + num_merged_bins
                quantized_bins[index] = sum(sliced_distribution[start:end])
            quantized_bins[-1] += sum(sliced_distribution[num_quantized_bins * num_merged_bins:])
```

say `num_half_quantized_bin = 64` and `zero_bin_index = 1024`, so loop will iterate from 64 to 1025.

When `i = 64`
```
sliced_distribution.size = 128
num_merged_bins = 1
```
that's pretty cool, because every element in quantized_bins just represent one element in sliced_distribution, no more  
outliers. In this case, quantized_bins is **exactly** what we want.

But if `sliced_distribution.size` is not integer times of  num_quantized_bins which default value is 128, what will happen ?


When `i = 127`
```
sliced_distribution.size = 254
num_merged_bins = 1
```

Even if we got a sliced distribution which contains 254 samples,  we still calculate the quantized_bins from the first 128 samples of  sliced distribution and just sum the left 126 samples to the last element of quantized_bins(quantized_bins[-1]). It  doesn't look like a good idea. I think this will cause more **accuracy problem**.


```
# start from 0 to 127
# end from 1 to 128
for index in range(num_quantized_bins):
      start = index * num_merged_bins
      end = start + num_merged_bins
      quantized_bins[index] = sum(sliced_distribution[start:end])
```

I have implemented a new version of `get_entropy_threshold` method base on https://github.com/Tencent/ncnn/blob/master/tools/quantize/ncnn2table.cpp (From L513). If you agree with me on  **accuracy problem**. I'm gladly to push commits to this PR.

I made sure that this wouldn’t take too much of your time.

Reply:
Thanks for providing another implementation of get_entropy_threshold.

Reply your question regarding "what will happen if sliced_distribution.size is not integer times of num_quantized_bins (default 128)?"
Yes, the situation you described will happen since our implementation of getting entropy threshold is the simple approach which only maintains histogram using integer. Therefore, for the case when ` i = 127`, you can imagine the KL divergence of this quantized distribution will be very large which makes it won't be chosen by np.argmin() later. When it comes to `i = 128`, 

    sliced_distribution.size = 256
    num_merged_bins = 2

This time the sliced distribution will be more evenly merged into the quantized bin. 
I quickly go over your implementation which seems provides more precise calculation in "merge sliced distribution into quantized bins". I will apply your approach to run several yolo models to see the accuracy results. Also, I was wondering what kind of models you are using? After applying your implementation, did you see accuracy improvement?

Reply:
@chilo-ms Thanks for your reply. It's a great pleasure for me to have this opportunity to share my idea on model quantization with you. But, I'm a little busy these days and some necessary test jobs should be done before push code to Github. Don't worry, I gonna do all of that on this weekend. 

Before that, I'm very interested in the way you do quantization test on yolo models. Can you share your test scripts or test repo (maybe) so I can got the accuracy results in the same way ?

Reply:
We have an e2e example here to guide how to use calibration-base quantization here and get mAP accuracy result.
https://github.com/microsoft/onnxruntime-inference-examples/blob/e2e-yolo-refine/quantization/object_detection/trt/yolov3/e2e_user_yolov3_example.py

You can also reference following repo and try to get accuracy parity with its result:
https://github.com/jkjung-avt/tensorrt_demos#yolov3

Reply:
@chilo-ms sorry for getting back to you so late, I'm on vacation recently,  you can take this commit as reference
https://github.com/xusworld/onnxruntime/commit/9234b692b263bc14012fd47733f1bc5594302d36